File: android/app/src/main/java/com/laneshadow/ui/atoms/LSTextField.kt
package com.laneshadow.ui.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.semantics.SemanticsPropertyKey
import androidx.compose.ui.semantics.SemanticsPropertyReceiver
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.editableText
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LaneShadowThemeValues
import com.laneshadow.theme.LocalLaneShadowTheme
import com.laneshadow.theme.generated.LaneShadowTheme.IconName

typealias LSIconAsset = IconName

val LSInputBorderColorKey = SemanticsPropertyKey<Color>("LSInputBorderColor")
val LSInputCornerRadiusKey = SemanticsPropertyKey<Dp>("LSInputCornerRadius")
val LSInputHorizontalPaddingKey = SemanticsPropertyKey<Dp>("LSInputHorizontalPadding")
val LSInputVerticalPaddingKey = SemanticsPropertyKey<Dp>("LSInputVerticalPadding")
val LSInputVisualStateKey = SemanticsPropertyKey<String>("LSInputVisualState")
val LSTextFieldSingleLineKey = SemanticsPropertyKey<Boolean>("LSTextFieldSingleLine")

internal var SemanticsPropertyReceiver.lsInputBorderColor by LSInputBorderColorKey
internal var SemanticsPropertyReceiver.lsInputCornerRadius by LSInputCornerRadiusKey
internal var SemanticsPropertyReceiver.lsInputHorizontalPadding by LSInputHorizontalPaddingKey
internal var SemanticsPropertyReceiver.lsInputVerticalPadding by LSInputVerticalPaddingKey
internal var SemanticsPropertyReceiver.lsInputVisualState by LSInputVisualStateKey
internal var SemanticsPropertyReceiver.lsTextFieldSingleLine by LSTextFieldSingleLineKey

internal val LaneShadowInputMinHeight = 48.dp
private val LaneShadowInputBorderWidth = 1.dp

@Composable
fun LSTextField(
    value: String,
    onValueChange: (String) -> Unit,
    state: InputState = InputState.Default,
    placeholder: String? = null,
    leadingIcon: LSIconAsset? = null,
    trailingIcon: LSIconAsset? = null,
    modifier: Modifier = Modifier,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
    keyboardActions: KeyboardActions = KeyboardActions.Default,
    visualTransformation: VisualTransformation = VisualTransformation.None,
) {
    val theme = LocalLaneShadowTheme.current
    val padding = resolveLSInputPadding(theme)
    val cornerRadius = resolveLSInputCornerRadius(theme)
    val contentColor = ContentColor.Primary.resolve(theme)
    val placeholderColor = ContentColor.Subtle.resolve(theme)
    var isFocused by remember { mutableStateOf(state == InputState.Focused) }
    val visualState = resolveLSInputVisualState(state, isFocused)
    val borderColor = resolveLSInputBorderColor(theme, visualState)
    val isEnabled = visualState != InputState.Disabled

    Surface(
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = LaneShadowInputMinHeight)
            .semantics(mergeDescendants = true) {
                contentDescription = placeholder ?: "Text field"
                editableText = AnnotatedString(value)
                stateDescription = visualState.name
                if (!isEnabled) {
                    disabled()
                }
                lsInputBorderColor = borderColor
                lsInputCornerRadius = cornerRadius
                lsInputHorizontalPadding = padding.calculateLeftPadding(LayoutDirection.Ltr)
                lsInputVerticalPadding = padding.calculateTopPadding()
                lsInputVisualState = visualState.name
                lsTextFieldSingleLine = true
            },
        color = theme.colors.input.default,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(cornerRadius),
        border = BorderStroke(LaneShadowInputBorderWidth, borderColor),
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(padding),
            horizontalArrangement = Arrangement.spacedBy(theme.space.sm),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            leadingIcon?.let { icon ->
                LSIcon(
                    name = icon,
                    size = IconSize.Sm,
                    color = IconColor.Content(
                        if (visualState == InputState.Error) ContentColor.Error else ContentColor.Secondary,
                    ),
                )
            }

            Box(
                modifier = Modifier.weight(1f),
                contentAlignment = Alignment.CenterStart,
            ) {
                if (value.isEmpty() && placeholder != null) {
                    LSText(
                        text = placeholder,
                        variant = TypographyVariant.Ui.Body.Md,
                        color = ContentColor.Subtle,
                    )
                }

                BasicTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier
                        .fillMaxWidth()
                        .onFocusChanged { isFocused = it.isFocused },
                    enabled = isEnabled,
                    singleLine = true,
                    textStyle = theme.typography.ui.body.md.copy(color = contentColor),
                    cursorBrush = SolidColor(if (visualState == InputState.Error) borderColor else placeholderColor),
                    keyboardOptions = keyboardOptions,
                    keyboardActions = keyboardActions,
                    visualTransformation = visualTransformation,
                )
            }

            trailingIcon?.let { icon ->
                LSIcon(
                    name = icon,
                    size = IconSize.Sm,
                    color = IconColor.Content(
                        if (visualState == InputState.Error) ContentColor.Error else ContentColor.Secondary,
                    ),
                    modifier = Modifier.size(theme.sizing.icon.sm),
                )
            }
        }
    }
}

internal fun resolveLSInputPadding(theme: LaneShadowThemeValues): PaddingValues =
    PaddingValues(horizontal = theme.space.md, vertical = theme.space.md)

internal fun resolveLSInputCornerRadius(theme: LaneShadowThemeValues): Dp = theme.radius.sm

internal fun resolveLSInputVisualState(
    requestedState: InputState,
    isFocused: Boolean,
): InputState =
    when {
        requestedState == InputState.Disabled -> InputState.Disabled
        requestedState == InputState.Error -> InputState.Error
        requestedState == InputState.Focused || isFocused -> InputState.Focused
        else -> InputState.Default
    }

internal fun resolveLSInputBorderColor(
    theme: LaneShadowThemeValues,
    state: InputState,
): Color =
    when (state) {
        InputState.Default -> theme.colors.border.default
        InputState.Focused -> theme.colors.border.focus ?: theme.colors.ring.default
        InputState.Error -> theme.colors.danger.default
        InputState.Disabled -> theme.colors.border.disabled ?: theme.colors.border.default
    }

File: ios/LaneShadow/Views/Atoms/LSTextField.swift
import LaneShadowTheme
import NativeTheme
import SwiftUI

public struct LSTextField: View {
    struct ResolvedTokens {
        let minHeight: CGFloat
        let horizontalPadding: CGFloat
        let cornerRadius: CGFloat
        let borderWidth: CGFloat
        let background: Color
        let border: Color
        let textColor: Color
        let placeholderColor: Color
        let helperColor: Color
        let iconColor: Color
        let iconContentColor: IconContentColor
        let iconSize: CGFloat
        let iconSpacing: CGFloat
        let textStyle: TypographyStyle
        let helperVariant: TypographyVariant
        let opacity: CGFloat
    }

    @Environment(\.theme) private var theme
    @FocusState private var isFocused: Bool

    @Binding private var value: String
    private let placeholder: String?
    private let state: InputState
    private let isSecureEntry: Bool
    private let leadingIcon: IconName?
    private let trailingIcon: IconName?
    private let leadingSymbolName: String?
    private let trailingSymbolName: String?
    private let helperText: String?
    private let inputAccessibilityIdentifier: String?

    public init(
        value: Binding<String>,
        placeholder: String? = nil,
        state: InputState = .default,
        isSecureEntry: Bool = false,
        leadingIcon: IconName? = nil,
        trailingIcon: IconName? = nil,
        leadingSymbolName: String? = nil,
        trailingSymbolName: String? = nil,
        helperText: String? = nil,
        inputAccessibilityIdentifier: String? = nil
    ) {
        _value = value
        self.placeholder = placeholder
        self.state = state
        self.isSecureEntry = isSecureEntry
        self.leadingIcon = leadingIcon
        self.trailingIcon = trailingIcon
        self.leadingSymbolName = leadingSymbolName
        self.trailingSymbolName = trailingSymbolName
        self.helperText = helperText
        self.inputAccessibilityIdentifier = inputAccessibilityIdentifier
    }

    public var body: some View {
        let tokens = Self.resolvedTokens(for: state, isFocused: isFocused, theme: theme)

        VStack(alignment: .leading, spacing: theme.space.xs) {
            HStack(spacing: tokens.iconSpacing) {
                if let leadingSymbolName {
                    LSIconSymbolIOS(
                        name: leadingSymbolName,
                        size: tokens.iconSize,
                        color: tokens.iconContentColor
                    )
                    .frame(width: tokens.iconSize, height: tokens.iconSize)
                } else if let leadingIcon {
                    LSIcon(name: leadingIcon, size: .sm, resolvedColorOverride: tokens.iconColor)
                        .frame(width: tokens.iconSize, height: tokens.iconSize)
                }

                inputField(tokens: tokens)

                if let trailingSymbolName {
                    LSIconSymbolIOS(
                        name: trailingSymbolName,
                        size: tokens.iconSize,
                        color: tokens.iconContentColor
                    )
                    .frame(width: tokens.iconSize, height: tokens.iconSize)
                } else if let trailingIcon {
                    LSIcon(name: trailingIcon, size: .sm, resolvedColorOverride: tokens.iconColor)
                        .frame(width: tokens.iconSize, height: tokens.iconSize)
                }
            }
            .padding(.horizontal, tokens.horizontalPadding)
            .frame(minHeight: tokens.minHeight)
            .background(tokens.background)
            .clipShape(RoundedRectangle(cornerRadius: tokens.cornerRadius))
            .overlay {
                RoundedRectangle(cornerRadius: tokens.cornerRadius)
                    .stroke(tokens.border, lineWidth: tokens.borderWidth)
            }
            .opacity(tokens.opacity)

            if let helperText {
                LSText(helperText, variant: tokens.helperVariant, color: .tertiary)
                    .accessibilityIdentifier("lstextfield-helper")
            }
        }
        .accessibilityIdentifier(inputAccessibilityIdentifier ?? "lstextfield")
    }

    static func resolvedTokens(
        for state: InputState,
        isFocused: Bool = false,
        theme: Theme
    ) -> ResolvedTokens {
        let visualState = visualState(for: state, isFocused: isFocused)
        let helperColor: Color = switch visualState {
        case .error:
            theme.colors.danger.default
        case .disabled:
            theme.colors.onSurface.disabled ?? theme.colors.onSurface.default.opacity(theme.opacity.disabled)
        case .default, .focused:
            ContentColor.tertiary.resolved(in: theme)
        }
        let iconColor: Color = switch visualState {
        case .error:
            theme.colors.danger.default
        case .disabled:
            theme.colors.onSurface.disabled ?? theme.colors.onSurface.default.opacity(theme.opacity.disabled)
        case .default, .focused:
            ContentColor.secondary.resolved(in: theme)
        }
        let iconContentColor: IconContentColor = switch visualState {
        case .disabled:
            .subtle
        case .default, .focused, .error:
            .secondary
        }

        return ResolvedTokens(
            minHeight: theme.control.minHeight,
            horizontalPadding: theme.space.md,
            cornerRadius: theme.radius.sm,
            borderWidth: theme.borderWidth.thin,
            background: backgroundColor(for: visualState, theme: theme),
            border: borderColor(for: visualState, theme: theme),
            textColor: textColor(for: visualState, theme: theme),
            placeholderColor: ContentColor.subtle.resolved(in: theme),
            helperColor: helperColor,
            iconColor: iconColor,
            iconContentColor: iconContentColor,
            iconSize: theme.iconSize.small,
            iconSpacing: theme.space.sm,
            textStyle: theme.type.body.lg,
            helperVariant: TypographyVariant.body.sm,
            opacity: visualState == .disabled ? theme.opacity.disabled : 1
        )
    }

    static func commitChange(current: String, proposed: String, state: InputState) -> String {
        state == .disabled ? current : proposed
    }

    private static func visualState(for state: InputState, isFocused: Bool) -> InputState {
        switch state {
        case .disabled:
            .disabled
        case .error:
            .error
        case .focused:
            .focused
        case .default:
            isFocused ? .focused : .default
        }
    }

    private static func backgroundColor(for state: InputState, theme: Theme) -> Color {
        switch state {
        case .disabled:
            theme.colors.input.disabled ?? theme.colors.surfaceVariant.default
        case .default, .focused, .error:
            theme.colors.input.default
        }
    }

    private static func borderColor(for state: InputState, theme: Theme) -> Color {
        switch state {
        case .error:
            theme.colors.danger.default
        case .focused:
            theme.colors.border.focus ?? theme.colors.primary.default
        case .disabled:
            theme.colors.border.disabled ?? theme.colors.border.default.opacity(theme.opacity.disabled)
        case .default:
            theme.colors.border.default
        }
    }

    private static func textColor(for state: InputState, theme: Theme) -> Color {
        switch state {
        case .disabled:
            theme.colors.onSurface.disabled ?? theme.colors.onSurface.default.opacity(theme.opacity.disabled)
        case .default, .focused, .error:
            ContentColor.primary.resolved(in: theme)
        }
    }

    private var editableBinding: Binding<String> {
        Binding(
            get: { value },
            set: { nextValue in
                value = Self.commitChange(current: value, proposed: nextValue, state: state)
            }
        )
    }

    private func prompt(tokens: ResolvedTokens) -> Text? {
        guard let placeholder else { return nil }
        return Text(placeholder)
            .font(tokens.textStyle.font)
            .foregroundStyle(tokens.placeholderColor)
    }

    @ViewBuilder
    private func inputField(tokens: ResolvedTokens) -> some View {
        if isSecureEntry {
            SecureField("", text: editableBinding, prompt: prompt(tokens: tokens))
                .font(tokens.textStyle.font)
                .foregroundStyle(tokens.textColor)
                .focused($isFocused)
                .disabled(state == .disabled)
                .optionalAccessibilityIdentifier(inputAccessibilityIdentifier)
        } else {
            TextField("", text: editableBinding, prompt: prompt(tokens: tokens))
                .font(tokens.textStyle.font)
                .foregroundStyle(tokens.textColor)
                .focused($isFocused)
                .disabled(state == .disabled)
                .optionalAccessibilityIdentifier(inputAccessibilityIdentifier)
        }
    }
}

private extension View {
    @ViewBuilder
    func optionalAccessibilityIdentifier(_ identifier: String?) -> some View {
        if let identifier {
            accessibilityIdentifier(identifier)
        } else {
            self
        }
    }
}
